src/libtls/tls_server.c

```diff
@@ -701,12 +992,87 @@ static status_t send_server_hello(private_tls_server_t *this,
 	/* NULL compression only */
 	writer->write_uint8(writer, 0);
 
+	if (version >= TLS_1_3)
+	{
+		extensions = bio_writer_create(32);
+
+		DBG2(DBG_TLS, "sending extension: %N",
+			 tls_extension_names, TLS_EXT_SUPPORTED_VERSIONS);
+		extensions->write_uint16(extensions, TLS_EXT_SUPPORTED_VERSIONS);
+		extensions->write_uint16(extensions, 2);
+		extensions->write_uint16(extensions, version);
+
+		if (this->dh)
+		{
+			tls_named_group_t selected_curve = this->requested_curve;
+
+			DBG2(DBG_TLS, "sending extension: %N",
+				 tls_extension_names, TLS_EXT_KEY_SHARE);
+			if (!this->dh->get_my_public_value(this->dh, &pub))
+			{
+				this->alert->add(this->alert, TLS_FATAL, TLS_INTERNAL_ERROR);
+				extensions->destroy(extensions);
+				return NEED_MORE;
+			}
+			extensions->write_uint16(extensions, TLS_EXT_KEY_SHARE);
+			key_share = bio_writer_create(pub.len + 6);
+			key_share->write_uint16(key_share, selected_curve);
+			if (selected_curve == TLS_CURVE25519 ||
+				selected_curve == TLS_CURVE448)
+			{
+				key_share->write_data16(key_share, pub);
+			}
+			else
+			{	/* classic format (see RFC 8446, section 4.2.8.2) */
+				key_share->write_uint16(key_share, pub.len + 1);
+				key_share->write_uint8(key_share, TLS_ANSI_UNCOMPRESSED);
+				key_share->write_data(key_share, pub);
+			}
+			extensions->write_data16(extensions, key_share->get_buf(key_share));
+			key_share->destroy(key_share);
+			free(pub.ptr);
```

Comment:
I am not sure about the namespace you mentioned. My suggestion would be a new file `tls_extension.c` which holds the implementation of `bool tls_write_key_share(bio_writer_t *writer, tls_named_group_t group, diffie_hellman_t *dh)` and both, `tls_server.c` and `tls_client.c` uses it via a forward-declaration. Do I miss something?

The implementation between client and server is different by one `key_share->wrap16(key_share)` line. Either I keep the current function declaration and use an additional new writer or I add a bool flag to the function declaration and make this line conditional.
I would suggest the former since I use this already in `TLS_EXT_SIGNATURE_ALGORITHMS`.

Reply:
No need for a new file. Just define the function in one of the two files and forward-declare it in the other. You have to namespace all non-static functions anyway (i.e. prefix the name with something like `tls_`) and until there are a lot of (related) helpers moving them to a separate file is really overkill.

> The implementation between client and server is different by one `key_share->wrap16(key_share)` line.

Yes, just let the helper write a single key share that you can either wrap or write directly to the extensions writer.

Reply:
Thank you for this clarification. I added an implementation in the latest commit.